### ISSUE TITLE
Passwordless Email Connection Unsupported

### ISSUE THREAD

**Issue by cs-jackb** [auth0/terraform-provider-auth0#174]
### Description

The `terraform-provider-auth0` implementation does not support Passwordless Email connection resources; the definitions for the `email` block are missing for connections with a strategy of `"email"`.  

A related, but substantively different, issue is filed here: https://github.com/auth0/terraform-provider-auth0/issues/31

### Affected Resource(s)

- auth0_connection

### Expected Behavior

Given the below connection from the Auth0 Management API `api/v2/connections` endpoint (as derived from the defaults for a new tenant), the `terraform-provider-auth0` should be able to successfully import this connection by ID as well as successfully plan and apply a resource that defines all aspects of this connection:
```json
{
    "id": "con_ABCDEFGHIJKLMNOP",
    "options": {
        "name": "email",
        "totp": {
            "length": 6,
            "time_step": 180
        },
        "email": {
            "body": "<html>Removed for brevity.</html>",
            "from": "{{ application.name }} <root@auth0.com>",
            "syntax": "liquid",
            "subject": "Welcome to {{ application.name }}"
        },
        "authParams": "",
        "disable_signup": false,
        "brute_force_protection": true
    },
    "strategy": "email",
    "name": "email",
    "is_domain_connection": false,
    "realms": [
        "email"
    ],
    "enabled_clients": [ ]
}
```
### Reproduction

Assuming a Passwordless Email connection is enabled in the tenant with ID `con_ABCDEFGHIJKLMNOP`, attempting to import the Passwordless Email connection results in an error: 
```bash
❯ terraform import auth0_connection.email con_ABCDEFGHIJKLMNOP

auth0_connection.email: Importing from ID "con_ABCDEFGHIJKLMNOP"...
auth0_connection.email: Import prepared!
  Prepared auth0_connection for import
auth0_connection.email: Refreshing state... [id=con_ABCDEFGHIJKLMNOP]
╷
│ Error: decoding response payload failed: json: cannot unmarshal string into Go struct field ConnectionOptionsEmail.authParams of type map[string]string
│ 
│ 
╵
```
Similarly, creating an `auth0_connection` resource like the below results in validation and plan failures for the `email` block: 
```tf
  resource "auth0_connection" "email" {
    enabled_clients      = [ ]
    is_domain_connection = false
    name                 = "email"
    realms               = [
      "email",
    ]
    strategy             = "email"
    options {
        totp {
          length                = 6
          time_step             = 180
        }
  
        email {
          body                  = "<html>Removed for brevity.</html>"
          from                  = "{{ application.name }} <root@auth0.com>"
          syntax                = "liquid"
          subject               = "Welcome to {{ application.name }}"
        }
  
        disable_signup          = false
        brute_force_protection  = true
    }
  }
```

```bash
❯ terraform plan                        
│ Error: Unsupported block type
│ 
│   on email.tf line 22, in resource "auth0_connection" "email":
│   22:       email {
│ 
│ Blocks of type "email" are not expected here.
╵
```
### Environment

The following version is defined for the `terraform-provider-auth0`: 
```tf
terraform {
    required_providers {
      auth0 = {
        source = "auth0/auth0"
        version = "~> 0.30.3"
      }
    }
  }
  
```
No other modules, plugins, or libraries are in use.

### Community Note
- Please vote on this issue by adding a :+1: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by cs-jackb** [comment]
I do see, in `resource_auth0_connection_test.go`, there's a test for email connections: 

```go
const testAccConnectionEmailConfigUpdate = `
resource "auth0_connection" "email" {
	name = "Acceptance-Test-Email-{{.random}}"
	is_domain_connection = false
	strategy = "email"
	options {
		disable_signup = false
		name = "Email OTP"
		from = "Magic Password <password@example.com>"
		subject = "Sign in!"
		syntax = "liquid"
		template = "<html><body><h1>Here's your password!</h1></body></html>"
		brute_force_protection = true
		totp {
			time_step = 360
			length = 4
		}
	}
}
`
```
This test indicates that the schema for the options has flatted the `email` object, and there's no logic in place to correctly parse the remote `email` object from the API result into the `options` map. So, it kinda seems like there's two options to maybe address this. 

### Option 1
Support the `email` block instead of the flattened structure. 

I'm assuming that it's just a matter of adding something like the below to the `resource_auth0_connection.go` under the `"options"` entry of the `Elem.Schema` in the `connectionSchema` var and updating the relevant unit tests: 

```go
// passwordless email options
"email": {
    Type:     schema.TypeList,
    Optional: true,
    MaxItems: 1,
    Elem: &schema.Resource{
        Schema: map[string]*schema.Schema{
            "body": {
                Type:     schema.TypeString,
                Optional: true,
                Description: "The content of the email your users are going to receive."
            },
            "from": {
                Type:     schema.TypeString,
                Optional: true,
            },
            "syntax": {
                Type:     schema.TypeString,
                Default:  "liquid",
                Optional: true,
		ValidateFunc: validation.StringInSlice([]string{
			"liquid",
				"markdown",
		}, true)
            },
            "subject": {
                Type:     schema.TypeString,
                Optional: true,
            },
        },
    },
    Description: "Customize the settings and options that will allow your users to use one-time password authentication via email messages.",
},
```

### Option 2
Update the read logic for the Connection to parse the `email` block from the API response to correctly flatten into the `options.from`, `options.subject`, `options.syntax`, and `options.template` entries.

**Comment by sergiughf** [comment]
Hey @cs-jackb thanks for opening this issue and for the thorough info attached as well! We'll look into this, but I don't have an exact ETA at the moment.

**Comment by cs-jackb** [comment]
@willvedd - thank you for reply. I am, however, confused still. What needs to change for the below command to succeed? Because right now, it is still impossible to successfully import an existing passwordless email connection. 

```bash 
> terraform import -allow-missing-config auth0_connection.email con_ABCDEFGHIJKLMNOP

auth0_connection.email: Importing from ID "con_ABCDEFGHIJKLMNOP"...
auth0_connection.email: Import prepared!
  Prepared auth0_connection for import
auth0_connection.email: Refreshing state... [id=con_ABCDEFGHIJKLMNOP]

╷
│ Error: decoding response payload failed: json: cannot unmarshal string into Go struct field ConnectionOptionsEmail.authParams of type map[string]string
│ 
│ 
╵
```

**Comment by cs-jackb** [comment]
The improved documentation in https://github.com/auth0/terraform-provider-auth0/pull/198 is appreciated, but the issue is definitely _not_ resolved. Please re-open.

**Comment by cs-jackb** [comment]
To maybe point you closer to the heart of the problem, the error message references an option property by the name of `authParams`, which shows up in the Management API v2/connections response, but is not present in the [flattenConnectionOptionsEmail](https://github.com/auth0/terraform-provider-auth0/blob/b8b5fb17bac0a4aceae6d355a7873902b5a835a9/auth0/structure_auth0_connection.go#L262) method you mention.

**Comment by sergiughf** [comment]
Hey @cs-jackb apologies for the confusion and appreciate your patience and help with debugging this, I'm reopening this. 

Indeed the `ConnectionOptionsEmail` struct defined in the [go-auth0 SDK](https://github.com/auth0/go-auth0/blob/be309b68ea3b9c120019fc9b6757825edfae309d/management/connection.go#L461) we use to leverage making those API calls has the following:

 ```
AuthParams map[string]string `json:"authParams,omitempty"
````

And we do not set that param yet in the terraform provider for a connection resource but we use that struct to unmarshal the payload. 

However the problem here is with the fact that that parameter is defined as a `map[string]string` but the API in your case is returning an empty string causing the json unmarshal to fail. 

As a workaround for now until we patch this, I believe that if you set the authParams to something like `{"scope": "openid profile"}` save it, and then remove it, the field should update to a correct `map[string]string` type that we can unmarshall and start responding correctly to a terraform import. 

We'll work on fixing this 💪🏻 Thanks again!

**Comment by sergiughf** [comment]
Hey @cs-jackb, we've patched this in v0.31.0. You should be able to import and manage SMS and Email Connections without any issues at the moment. As such we'll be closing this issue down:) Let us know if anything else pops up and feel free to reopen this if needed.

**Comment by cs-jackb** [comment]
Much appreciated, @sergiughf!
